File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function(s, k) {
    let l=0, r=0, res=0
    const map = new Map()
    let maxF = 0
    while(r < s.length) {
        map.set(s[r], (map.get(s[r]) || 0) + 1)
        maxF = Math.max(maxF, map.get(s[r]))
        if(r-l+1 - maxF <= k) {
            res = Math.max(res, r-l+1)
            r++
        } else {
            map.set(s[l], map.get(s[l]) - 1)
            map.set(s[r], (map.get(s[r]) || 0) - 1)
            l++
        }
    }
    return res 
};

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    int getMapFreq(Map<Character, Integer> map) {
        int max = 0;
        for(int val: map.values()) {
            max = Math.max(max, val);
        }
        return max;
    }

    public int characterReplacement(String s, int k) {
        int l=0, r=0;
        Map<Character, Integer> map = new HashMap();
        int longestWindowSize = 0;
        while(r < s.length()) {
            char ch = s.charAt(r);
            map.put(ch, map.getOrDefault(ch, 0) + 1);
            int maxMapFreq = getMapFreq(map);
            // System.out.println("ff: "+ maxMapFreq);
            if((r-l+1 - maxMapFreq) <= k) {
                longestWindowSize = Math.max(longestWindowSize, r-l+1);
            } else {
                char charAtL = s.charAt(l);
                map.put(charAtL, map.getOrDefault(charAtL, 0) - 1);
                l++;
            }
            r++;
        }
        return longestWindowSize;
    }
}